### DISCUSSION TITLE
FreeRTOS porting

### DISCUSSION THREAD

**Discussion by Gonta01** [nasa/fprime#2834] [Q&A] [answered]
Hi.
During freeRTOS porting the following problem occurred:
```
/home/user/projects/fprime_sbc/fprime/Os/TaskId.hpp:24:9: error: 'TaskIdRepr' does not name a type; did you mean 'TaskId'?
   24 |         TaskIdRepr getRepr() const;
      |         ^~~~~~~~~~
      |         TaskId
/home/user/projects/fprime_sbc/fprime/Os/TaskId.hpp:26:9: error: 'TaskIdRepr' does not name a type; did you mean 'TaskId'?
   26 |         TaskIdRepr id;
      |         ^~~~~~~~~~
      |         TaskId
```

The reason is clear:
In file fprime/Os/TaskIdRepr.hpp there are lines:
```
namespace Os {
#if defined(TGT_OS_TYPE_VXWORKS) || (FW_BAREMETAL_SCHEDULER == 1)
    typedef int TaskIdRepr;
#elif defined(TGT_OS_TYPE_LINUX) || defined(TGT_OS_TYPE_DARWIN)
    typedef pthread_t TaskIdRepr;
#endif
}
```
Since neither of defines is present, as a result there is no definition of TaskIdRepr.

What is the best solution without modifying third party code (fprime)?

**Comment by Gonta01** [comment]
Also is it in plans to add TGT_OS_TYPE_FREERTOS?
So that the code will look like this:
```
namespace Os {
#if defined(TGT_OS_TYPE_VXWORKS) || defined(TGT_OS_TYPE_FREERTOS) || (FW_BAREMETAL_SCHEDULER == 1)
    typedef int TaskIdRepr;
#elif defined(TGT_OS_TYPE_LINUX) || defined(TGT_OS_TYPE_DARWIN)
    typedef pthread_t TaskIdRepr;
#endif
}
```

**Reply by SterlingPeet** [reply]
The Arduino port leverages the BAREMETAL_SCHEDULER, so its not an exact solution to your problem; however you can use similar techniques where you provide that define in you library and then compile the libOs and libFreeRTOSOs as co-dependencies:

https://github.com/fprime-community/fprime-arduino/tree/main/Arduino/Os

Specifically the Arduino Os CMakeLists.txt file goes through and removes conflicting cpp files and adds appropriate cpp files to compile for the platform specific Os implementation.

**Reply by Gonta01** [reply]
@SterlingPeet, where exactly you suggest to define  TaskIdRepr? In cpp or hpp file of my library? How it will become visible in TaskId.hpp file?

"the Arduino Os CMakeLists.txt file goes through and removes conflicting cpp files" - could you please show these lines?

**Reply by SterlingPeet** [reply]
My apologies about the the "removing conflicting cpp files" comment, I was not looking directly at the code.  That functionality is available in cmake if you need it the F Prime tooling has an example here:

https://github.com/nasa/fprime/blob/dd463e1f6d4c64e884ec231e9e3fc7e7c9d313f0/Os/CMakeLists.txt#L83-L90

However, you may not actually need to do that.  Within the Arduino CMakeLists.txt in the folder linked above is the following nugget for combining your FreeRTOS OS library with the F Prime Os internal library:

https://github.com/fprime-community/fprime-arduino/blob/ffd162f88411a28eb5dad669bd49580ae3381084/Arduino/Os/CMakeLists.txt#L56-L67

(reproduced here, because its a different repo)

```cmake
# Dependency Melding:
#
# In short, this Os layer depends on the framework's Os layer and vise-versa. Effectively it is one
# library built in two halves, the framework base, and this add on. Here we make the link pointers
# depend on one another forming a (weak) circular dependency. This is allowed in CMake 3.13 and
# newer.
####
if (FPRIME_ARDUINO)
    add_dependencies(Os Arduino_Os)
    target_link_libraries(Os PRIVATE Arduino_Os)
    target_link_libraries(Arduino_Os PUBLIC Os)
endif()
```

The key to understand here is that you do _NOT_ need to put all of the function definitions and prototypes for the library in the same file.  The definitions just have to exist in one (and only one for each signature) of the files included in the cmake-managed compiler call to compile the object file.  So, for you missing hpp definition, you need to include a definition for ``TaskIdRepr`` in your FreeRTOS-specific Os cmake target.

To get ``Task.hpp`` to include it, it needs to be in the correct place for F Prime's cmake tooling to find it.  This means it should be located in ``Os/TaskIdRepr.hpp`` in your project or library, and cmake will sure that it can be found and included.

There is a certainly a bit of complexity in this process, and I would have to work through it to confirm all of the details.  Hopefully that gets you a path forward, and we can massage more details as they come up.

**Reply by Gonta01** [reply, marked as answer]
The definition of TaskIdRepr was written in PlatformTypes.h, similarly as it was explained by @ethancheez in [issue ](https://github.com/fprime-community/fprime-zephyr-led-blinker/issues/3#issuecomment-2289242612)

**Reply by SterlingPeet** [reply]
yep, @ethancheez 's solution is more directly applicable to your case.

**Reply by Gonta01** [reply]
@SterlingPeet your feedback also appreciated, it helped to clarify some parts of the project. Thanks.
